python/ray/data/datasource/text_datasource.py

```diff
@@ -0,0 +1,37 @@
+from typing import TYPE_CHECKING, List, Union, Tuple
+
+from ray.data.datasource.binary_datasource import BinaryDatasource
+from ray.util.annotations import PublicAPI
+
+
+if TYPE_CHECKING:
+    import pyarrow
+
+
+@PublicAPI
+class TextDatasource(BinaryDatasource):
+    """Text datasource, for reading and writing text files."""
+
+    _COLUMN_NAME = "text"
+
+    def _read_file(
+        self, f: "pyarrow.NativeFile", path: str, **reader_args
+    ) -> List[str]:
+        block = super()._read_file(f, path, **reader_args)
+        assert len(block) == 1
+        data = block[0]
+
+        lines = data.decode(reader_args["encoding"]).split("\n")
+        if reader_args["drop_empty_lines"]:
+            lines = [line for line in lines if line.strip() != ""]
+        return lines
+
+    def _convert_block_to_tabular_block(
+        self, block: List[Union[bytes, Tuple[bytes, str]]]
+    ) -> "pyarrow.Table":
+        import pyarrow as pa
+
+        return pa.Table.from_pydict({self._COLUMN_NAME: block})
```

Comment:
Nit: Using the `pa.table()` API, when possible, is generally advised.
```suggestion
        return pa.table({self._COLUMN_NAME: block})
```

Reply:
Addressed. 

ooc why is `pa.table` advised?

Reply:
`pa.table()` is pyarrow's canonical "create a table" API, where they centralize a lot of "take what the user gives us and build a well-defined table" logic. This is similar to `np.array()` for NumPy and `pd.DataFrame()` for pandas. We've tried to always use these idiomatic builder APIs in tests when creating library-specific abstractions from Python objects so we're exposing the smallest possible library API surface to the tests (i.e. test developers only have to be aware of the semantics for that builder API, not `pa.Table.from_pydict()`, `pa.Table.from_pylist()`, `pa.Table.from_pandas()`, etc.).

Reply:
An argument could be made for always using the narrowest API possible, e.g. using `pa.Table.from_pydict()` when building an Arrow table from a dict, when the caller knows that narrow API suffices, but for the sake of tests, I think that the improved test dev experience wins out. Lmk if you disagree!

Reply:
Makes sense to me. Thanks for the context.